File: bukkit/universalelectricity/Vector2.java
package universalelectricity;

import net.minecraft.server.MathHelper;


/**
 * The Class Vector2.
 */
public class Vector2
{
    
    /** The x. */
    public double x;
    
    /** The y. */
    public double y;

    /**
     * Instantiates a new vector2.
     */
    public Vector2()
    {
        this(0, 0);
    }

    /**
     * Instantiates a new vector2.
     *
     * @param i the i
     * @param j the j
     */
    public Vector2(int i, int j)
    {
        x = i;
        y = j;
    }

    /**
     * Instantiates a new vector2.
     *
     * @param d the d
     * @param d1 the d1
     */
    public Vector2(double d, double d1)
    {
        x = d;
        y = d1;
    }

    /**
     * Int x.
     *
     * @return the int
     */
    public int intX()
    {
        return (int)Math.floor(x);
    }

    /**
     * Int y.
     *
     * @return the int
     */
    public int intY()
    {
        return (int)Math.floor(y);
    }

    /**
     * Checks if is point in region.
     *
     * @param vector3 the vector3
     * @param vector3_1 the vector3_1
     * @param vector3_2 the vector3_2
     * @return true, if is point in region
     */
    public static boolean isPointInRegion(Vector2 vector2, Vector2 vector2_1, Vector2 vector2_2)
    {
        return vector2.x > vector2_1.x && vector2.x < vector2_2.x && vector2.y > vector2_1.y && vector2.y < vector2_2.y;
    }

    /**
     * Distance.
     *
     * @param vector2 the vector2
     * @param vector2_1 the vector2_1
     * @return the double
     */
    public static double distance(Vector2 vector2, Vector2 vector2_1)
    {
        double d = vector2.x - vector2_1.x;
        double d1 = vector2.y - vector2_1.y;
        return (double)MathHelper.sqrt(d * d + d1 * d1);
    }

    /**
     * Slope.
     *
     * @param vector2 the vector2
     * @param vector2_1 the vector2_1
     * @return the double
     */
    public static double slope(Vector2 vector2, Vector2 vector2_1)
    {
        double d = vector2.x - vector2_1.x;
        double d1 = vector2.y - vector2_1.y;
        return d1 / d;
    }

    /**
     * Adds the.
     *
     * @param vector2 the vector2
     */
    public void add(Vector2 vector2)
    {
        x += vector2.x;
        y += vector2.y;
    }

    /**
     * Adds the.
     *
     * @param d the d
     */
    public void add(double d)
    {
        x += d;
        y += d;
    }

    /**
     * Round.
     *
     * @return the vector2
     */
    public Vector2 round()
    {
        return new Vector2(Math.round(x), Math.round(y));
    }

    /**
     * Floor.
     *
     * @return the vector2
     */
    public Vector2 floor()
    {
        return new Vector2(Math.floor(x), Math.floor(y));
    }

    /**
     * Output.
     *
     * @return the string
     */
    public String output()
    {
        return (new StringBuilder()).append("Vector2: ").append(x).append(",").append(y).toString();
    }

    /**
     * Op.
     */
    public void op()
    {
        System.out.println(output());
    }
}

File: src/main/scala/universalelectricity/core/vector/Vector2.scala
package universalelectricity.core.vector

import java.lang.Double.doubleToLongBits
import net.minecraft.nbt.NBTTagCompound

/**
 * Vector2 Class is used for defining objects in a 2D space.
 *
 * @author Calclavia
 */
class Vector2 extends Cloneable
{
  var x = 0D
  var y = 0D

  def this(x: Double, y: Double)
  {
    this()
    this.x = x
    this.y = y
  }

  def x(amount: Double)
  {
    x = amount
  }

  def y(amount: Double)
  {
    y = amount
  }

  def xi = x.toInt

  def yi = y.toInt

  def xf = x.toFloat

  def yf = y.toFloat

  def round: Vector2 = new Vector2(Math.round(x), Math.round(y))

  def ceil: Vector2 = new Vector2(Math.ceil(x), Math.ceil(y))

  def floor: Vector2 = new Vector2(Math.floor(x), Math.floor(y))

  def max(other: Vector2): Vector2 = new Vector2(Math.max(x, other.x), Math.max(y, other.y))

  def min(other: Vector2): Vector2 = new Vector2(Math.min(x, other.x), Math.min(y, other.y))

  override def clone: Vector2 = return new Vector2(this.x, this.y)

  /**
   * Operations
   */
  def +(amount: Double): Vector2 = new Vector2(x + amount, y + amount)

  def +(amount: Vector2): Vector2 = new Vector2(x + amount.x, y + amount.y)

  def -(amount: Double): Vector2 = this + -amount

  def -(amount: Vector2): Vector2 = this + (amount * -1)

  def *(amount: Double): Vector2 = new Vector2(x * amount, y * amount)

  def /(amount: Double): Vector2 = new Vector2(x / amount, y / amount)

  def +=(amount: Double): Vector2 =
  {
    x += amount
    y += amount
    return this
  }

  def +=(amount: Vector2): Vector2 =
  {
    x += amount.x
    y += amount.y
    return this
  }

  def -=(amount: Double): Vector2 =
  {
    this += -amount
    return this
  }

  def -=(amount: Vector2): Vector2 =
  {
    this += (amount * -1)
    return this
  }

  def *=(amount: Double): Vector2 =
  {
    x *= amount
    y *= amount
    return this
  }

  def *=(amount: Vector2): Vector2 =
  {
    x *= amount.x
    y *= amount.y
    return this
  }

  def /=(amount: Double): Vector2 =
  {
    this *= (1 / amount)
    return this
  }


  def /=(amount: Vector2): Vector2 =
  {
    x /= amount.x
    y /= amount.y
    return this
  }

  /**
   * Magnitude
   */
  def dot(other: Vector2) = $(other)

  def $(other: Vector2) = x * other.x + y * other.y

  def magnitudeSquared = this $ this;

  def magnitude = Math.sqrt(magnitudeSquared)

  def normalize = this / magnitude

  def distance(other: Vector3) = (other - this).magnitude

  def midpoint(pos: Vector2): Vector2 = new Vector2((x + pos.x) / 2, (y + pos.y) / 2)

  def slope(other: Vector2) = (y - other.y) / (x - other.x)

  def writeToNBT(nbt: NBTTagCompound): NBTTagCompound =
  {
    nbt.setDouble("x", x)
    nbt.setDouble("y", y)
    return nbt
  }

  override def hashCode: Int =
  {
    val x = doubleToLongBits(this.x)
    val y = doubleToLongBits(this.y)
    return 31 * (x ^ (x >>> 32)).toInt + (y ^ (y >>> 32)).toInt
  }

  override def equals(o: Any): Boolean =
  {
    if (o.isInstanceOf[Vector2])
    {
      val vector: Vector2 = o.asInstanceOf[Vector2]
      return this.x == vector.x && this.y == vector.y
    }
    return false
  }

  override def toString = "Vector2 [" + this.x + "," + this.y + "]"
}